phpBB/includes/acp/acp_main.php

```diff
@@ -679,17 +679,19 @@ function main($id, $mode)
 			}
 		}
 
-		if (!defined('PHPBB_DISABLE_CONFIG_CHECK') && file_exists($phpbb_root_path . 'config.' . $phpEx) && $phpbb_filesystem->is_writable($phpbb_root_path . 'config.' . $phpEx))
+		if (!defined('PHPBB_DISABLE_CONFIG_CHECK'))
 		{
-			// World-Writable? (000x)
-			$template->assign_var('S_WRITABLE_CONFIG', (bool) (@fileperms($phpbb_root_path . 'config.' . $phpEx) & 0x0002));
+			$phpbb_config_file = $phpbb_root_path . 'config.' . $phpEx;
+
+			$template->assign_var('S_WRITABLE_CONFIG', (bool) $phpbb_filesystem->is_writable($phpbb_config_file) || (@fileperms($phpbb_config_file) & 0x0002));
```

Comment:
The file permission mask is defined in binary as:
- first 3 bits: world (w)
- second 3 bits: group (g)
- third 3 bits: user (u)

So essentially (first bit is right): uuu ggg www
Each of those 3 bit groups are defining read, write, execute. Therefore 000 000 010 will define that it's world writable, no matter what the other bytes are defining. This evaluates to 0x0002 and hence the previous check did in fact confirm whether the file is world writable. Your change will already evaluate to true no matter if it's writable to the current user, the current group, or world writable. As a result, this complains about config.php being world writable even though it has a permission mask of 600.

Reply:
Wouldn't it make more sense to check if the file is readable before checking its file permissions? Judging by the ticket that could already solve this issue?

Reply:
What would be your proposal then, example?

Reply:
Let's maybe take a step back. The initial issue is that the `@fileperm()` fails cause the file is not readable? At least that's what I'm deducting from the ticket. Therefore, you opted to instead check if the file is writable which unfortunately also says it's writable when the current user is able to write to it.
However, this check should only check for file permissions for "world". Therefore, I think it makes sense to first check if the file can be read before using `fileperm()`. That way we'll ensure that it's both readable and that `fileperm()` can be checked, i.e.:
```suggestion
			$template->assign_var('S_WRITABLE_CONFIG', (bool) $phpbb_filesystem->is_readable($phpbb_config_file) && (@fileperms($phpbb_config_file) & 0x0002));
```

Reply:
IMHO there is no need to check if the `config.php` file exists (and is readable) unless I am missing something.
If it didn't you could never get to the code in question because the forum would probably not work, for sure.

The origin of the issue as per the ticket it was in the function `phpbb_is_writable` of the `phpbb/filesystem`
![immagine](https://user-images.githubusercontent.com/480857/103426057-ed29c200-4bb6-11eb-8c95-d1354f24b4d4.png)

```
				$handle = @fopen($file, 'c');
				if (is_resource($handle))
				{
				return $handle->isWritable();
					fclose($handle);
					return true;
				}
```
That is being fixed / rewritten in this PR.

The cleanup of the ACP / tpl switch is a bonus, after a couple of more tests I can now push a new commit which should satisfy the logic in its whole, tested.